clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp

```diff
@@ -1,51 +1,47 @@
-// RUN: %clang_cc1 -fsycl -fsycl-is-device -Wno-return-type -fcxx-exceptions -fsyntax-only -ast-dump -verify -pedantic %s | FileCheck %s
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -Wno-return-type -Wno-sycl-2017-compat -fcxx-exceptions -fsyntax-only -ast-dump -verify -pedantic %s | FileCheck %s
+
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
 
 struct FuncObj {
   //expected-warning@+2 {{attribute 'intelfpga::no_global_work_offset' is deprecated}}
   //expected-note@+1 {{did you mean to use 'intel::no_global_work_offset' instead?}}
-  [[intelfpga::no_global_work_offset]] void operator()() {}
+  [[intelfpga::no_global_work_offset]] void operator()() const {}
 };
 
-template <typename name, typename Func>
-void kernel(Func kernelFunc) {
-  kernelFunc();
-}
-
 int main() {
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}Enabled
-  kernel<class test_kernel1>([]() {
-    FuncObj();
-  });
-
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr
-  // CHECK-NOT: Enabled
-  kernel<class test_kernel2>(
-      []() [[intel::no_global_work_offset(0)]]{});
-
-  // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}Enabled
-  // expected-warning@+2{{'no_global_work_offset' attribute should be 0 or 1. Adjusted to 1}}
-  kernel<class test_kernel3>(
-      []() [[intel::no_global_work_offset(42)]]{});
-
-  // expected-error@+2{{'no_global_work_offset' attribute requires a non-negative integral compile time constant expression}}
-  kernel<class test_kernel4>(
-      []() [[intel::no_global_work_offset(-1)]]{});
-
-  // expected-error@+2{{'no_global_work_offset' attribute requires parameter 0 to be an integer constant}}
-  kernel<class test_kernel5>(
-      []() [[intel::no_global_work_offset("foo")]]{});
-
-  kernel<class test_kernel6>([]() {
-    // expected-error@+1{{'no_global_work_offset' attribute only applies to functions}}
-    [[intel::no_global_work_offset(1)]] int a;
+  q.submit([&](handler &h) {
+    // CHECK:       SYCLIntelNoGlobalWorkOffsetAttr {{.*}}
+    // CHECK-NEXT:  IntegerLiteral{{.*}}1{{$}}
+    h.single_task<class test_kernel1>(FuncObj());
+
+    // CHECK: SYCLIntelNoGlobalWorkOffsetAttr {{.*}}
+    // CHECK-NEXT:  IntegerLiteral{{.*}}0{{$}}
+    h.single_task<class test_kernel2>(
+        []() [[intel::no_global_work_offset(0)]]{});
+
+    // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}
+    h.single_task<class test_kernel3>(
+        []() [[intel::no_global_work_offset(42)]]{});
+
+    // CHECK: SYCLIntelNoGlobalWorkOffsetAttr{{.*}}
+    h.single_task<class test_kernel4>(
+        []() [[intel::no_global_work_offset(-1)]]{});
```

Comment:
Please correct me if I've misunderstood this, but any integer constant is now allowed? Does -1 default to 1 here? Can you please add CHECKs for Integer literals as well in this test?

Reply:
The attribute takes a Boolean expression. Any nonzero number should be treated as true.
So -1 or 42 is a valid Boolean expression that means true. Zero should be treated as false.

This will make the codes more complicated if we correct the value here (default to 1) 
when creating the semantic attribute. 

Please see the discussion here as a reference.
https://github.com/intel/llvm/pull/2839#discussion_r543398214

After discussion with @AaronBallman, it was decided that if there is a reason, we can make the changes and could correct the value (makes -1 or 42 to default 1).  We will currently leave this now and would not give any diagnostic here.

Reply:
Oh I see. Thanks for explaining.